### PR TITLE
1.x: Add maxConcurrent parameter to concatMapEager

### PR DESCRIPTION
**Pull request by adi1133**
Relevant issue #3710

### CODE REVIEW

File: src/main/java/rx/Observable.java [hunk 1]
```diff
@@ -5132,7 +5132,39 @@ public final void onNext(T args) {
...
+     * @warn javadoc fails to describe the return value
+     * @since (if this graduates from Experimental/Beta to supported, replace this parenthetical with the release number)
+     */
+
+    @Experimental
+    public final <R> Observable<R> concatMapEager(Func1<? super T, ? extends Observable<? extends R>> mapper, int capacityHint, int maxConcurrent) {
+        if (capacityHint < 1) {
+            throw new IllegalArgumentException("capacityHint > 0 required but it was " + capacityHint);
+        }
+        return lift(new OperatorEagerConcatMap<T, R>(mapper, capacityHint, maxConcurrent));
```
Comment (akarnokd) [inline src/main/java/rx/Observable.java]: Please validate `maxConcurrent` as well.

File: src/main/java/rx/internal/operators/OperatorEagerConcatMap.java [hunk 2]
```diff
@@ -220,6 +223,7 @@ void drain() {
                             if (empty) {
                                 synchronized (subscribers) {
                                     subscribers.poll();
+                                    request(1);
```
Comment (akarnokd) [inline src/main/java/rx/internal/operators/OperatorEagerConcatMap.java]: Please call `request(1)` outside the synchronized block and preferrably after `innerDone = true`.

**Comment by akarnokd** [comment]
Will you update this PR?

**Comment by adi1133** [comment]
Yes, on it !

**Comment by adi1133** [comment]
I submitted a commit with your proposed changes, sorry for the delay. 
Feel free to give me more comments/changes, also I can squash if needed.

**Comment by akarnokd** [comment]
:+1:

Don't squash this time. Github has now a new feature for the merge button where it offers squashing. I'd like to test if it really squashes everything and the main just gets 1 commit.

File: src/main/java/rx/Observable.java [hunk 3]
```diff
@@ -5132,7 +5132,42 @@ public final void onNext(T args) {
...
+     * @param <R> the value type
+     * @param mapper the function that maps a sequence of values into a sequence of Observables that will be
+     *               eagerly concatenated
+     * @param capacityHint hints about the number of expected source sequence values
+     * @param maxConcurrent the maximum number of concurrent subscribed observables
+     * @return
+     * @warn javadoc fails to describe the return value
+     * @since (if this graduates from Experimental/Beta to supported, replace this parenthetical with the release number)
+     */
+
```
Comment (artem-zinnatullin) [inline src/main/java/rx/Observable.java]: nit: empty line

File: src/test/java/rx/internal/operators/OperatorEagerConcatMapTest.java [hunk 4]
```diff
@@ -405,9 +407,34 @@ public void testInnerNull() {
...
+        ts.assertValueCount(100);
+        ts.assertCompleted();
+
+        Assert.assertEquals(5, (long) requests.get(0));
+        Assert.assertEquals(1, (long) requests.get(1));
+        Assert.assertEquals(1, (long) requests.get(2));
+        Assert.assertEquals(1, (long) requests.get(3));
+        Assert.assertEquals(1, (long) requests.get(4));
+        Assert.assertEquals(1, (long) requests.get(5));
+
```
Comment (artem-zinnatullin) [inline src/test/java/rx/internal/operators/OperatorEagerConcatMapTest.java]: nit: 2 empty lines for no reason

File: src/test/java/rx/internal/operators/OperatorEagerConcatMapTest.java [hunk 5]
```diff
@@ -405,9 +407,34 @@ public void testInnerNull() {
...
-        
+
         ts.assertNoErrors();
         ts.assertCompleted();
         ts.assertValue(null);
     }
+
+
+    @Test
+    public void testMaxConcurrent5() {
```
Comment (artem-zinnatullin) [inline src/test/java/rx/internal/operators/OperatorEagerConcatMapTest.java]: nit: test prefix is not needed

File: src/main/java/rx/Observable.java [hunk 6]
```diff
@@ -5132,7 +5132,42 @@ public final void onNext(T args) {
...
+    @Experimental
+    public final <R> Observable<R> concatMapEager(Func1<? super T, ? extends Observable<? extends R>> mapper, int capacityHint, int maxConcurrent) {
+        if (capacityHint < 1) {
```
Comment (artem-zinnatullin) [inline src/main/java/rx/Observable.java]: Would be great to cover this with tests too

File: src/main/java/rx/Observable.java [hunk 7]
```diff
@@ -5132,7 +5132,42 @@ public final void onNext(T args) {
...
+            throw new IllegalArgumentException("capacityHint > 0 required but it was " + capacityHint);
+        }
+        if (maxConcurrent < 1) {
```
Comment (artem-zinnatullin) [inline src/main/java/rx/Observable.java]: Same as above

**Comment by artem-zinnatullin** [comment]
Nothing critical, :+1: 

// @akarnokd squashing on GitHub works, and it works especially great when author of the PR just commits on top of previous commits because GitHub now remembers what code you've checked and what not and you can continue reviewing from previous point (doesn't work for me when PR is always squashed).

File: src/main/java/rx/Observable.java (hunk 6 above)
Comment (adi1133) [inline src/main/java/rx/Observable.java]: This is already covered in `testInvalidCapacityHint()`

**Comment by adi1133** [comment]
I have updated the PR. Do PR commits trigger github notifications ? (Should I comment after I do an update ? )

**Comment by akarnokd** [comment]
Unfortunately, no notification.

**Comment by stevegury** [comment]
👍

**Pull request merged**